clients/src/test/java/org/apache/kafka/clients/producer/internals/TransactionManagerTest.java

```diff
@@ -1385,6 +1386,119 @@ public void testRaiseErrorWhenNoPartitionsPendingOnDrain() throws InterruptedExc
         assertTrue(drainedBatches.get(node1.id()).isEmpty());
     }
 
+    @Test
+    public void testTransitionToAbortableErrorOnBatchExpiry() throws InterruptedException, ExecutionException {
+        final long pid = 13131L;
+        final short epoch = 1;
+
+        doInitTransactions(pid, epoch);
+
+        transactionManager.beginTransaction();
+        transactionManager.maybeAddPartitionToTransaction(tp0);
+
+        Future<RecordMetadata> responseFuture = accumulator.append(tp0, time.milliseconds(), "key".getBytes(),
+                "value".getBytes(), Record.EMPTY_HEADERS, null, MAX_BLOCK_TIMEOUT).future;
+
+        assertFalse(responseFuture.isDone());
+
+        prepareAddPartitionsToTxnResponse(Errors.NONE, tp0, epoch, pid);
+
+        assertFalse(transactionManager.transactionContainsPartition(tp0));
+        assertFalse(transactionManager.sendToPartitionAllowed(tp0));
+        sender.run(time.milliseconds());  // send addPartitions.
+        // Check that only addPartitions was sent.
+        assertTrue(transactionManager.transactionContainsPartition(tp0));
+        assertTrue(transactionManager.sendToPartitionAllowed(tp0));
+        assertFalse(responseFuture.isDone());
+
+        // Sleep 10 seconds to make sure that the batches in the queue would be expired if they can't be drained.
+        time.sleep(10000);
+        // Disconnect the target node for the pending produce request. This will ensure that sender will try to
+        // expire the batch.
+        Node clusterNode = this.cluster.nodes().get(0);
+        client.disconnect(clusterNode.idString());
+        client.blackout(clusterNode, 100);
+
+        sender.run(time.milliseconds());  // We should try to flush the produce, but expire it instead without sending anything.
+        assertTrue(responseFuture.isDone());
+
+        try {
+            // make sure the produce was expired.
+            responseFuture.get();
+            fail("Expected to get a TimeoutException since the queued ProducerBatch should have been expired");
+        } catch (ExecutionException e) {
+            assertTrue(e.getCause() instanceof  TimeoutException);
+        }
+        assertTrue(transactionManager.hasAbortableError());
+    }
+
+    @Test
+    public void testDropCommitOnBatchExpiry() throws InterruptedException, ExecutionException {
```

Comment:
There is a race condition I'm thinking about. Consider this sequence:

1. User calls commit.
2. One of the batches fails with an abortable error.
3. User calls abort.

Now, after we transition to the ABORTABLE_ERROR state, the commit request will still be in the pending request queue. We expect that it will be cleared in `nextRequestHandler()` on the next call to `maybeSendTransactionalRequest`. However, that happens asynchronously, so it seems possible that step 3 happens first and we transition to ABORTING_TRANSACTION. In that case, `nextRequestHandler()` will not cancel the commit request.

To address this, it seems like we need to ensure that a transition to an error state clears any pending EndTxn request. Does that make sense?